fix(y-mongodb.js): Prevent service disruption due to failed data updates

lukenc opened the pull request:
- In `src/y-mongodb.js`:
  - Enhanced the error handling for `Y.applyUpdate()` failures: If an exception occurs while attempting to apply any update, not only log a warning but also trigger a safety mechanism to halt the current document processing flow. This prevents potential data inconsistencies or more severe system faults.
  - Added an urgent notification mechanism when critical update application failures are detected, allowing the operations team to respond promptly.
  - Updated logging to provide more detailed information, including but not limited to the type of error, the name of the affected document, and specific update content, aiding in quick diagnosis of the issue.

This fix aims to enhance system robustness and reliability, particularly in handling anomalies from external data sources, preventing localized errors from escalating into widespread service outages.

lukenc committed:
fix(y-mongodb.js): Prevent service disruption due to failed data updates

- In `src/y-mongodb.js`:
  - Enhanced the error handling for `Y.applyUpdate()` failures: If an exception occurs while attempting to apply any update, not only log a warning but also trigger a safety mechanism to halt the current document processing flow. This prevents potential data inconsistencies or more severe system faults.
  - Added an urgent notification mechanism when critical update application failures are detected, allowing the operations team to respond promptly.
  - Updated logging to provide more detailed information, including but not limited to the type of error, the name of the affected document, and specific update content, aiding in quick diagnosis of the issue.

This fix aims to enhance system robustness and reliability, particularly in handling anomalies from external data sources, preventing localized errors from escalating into widespread service outages.

lukenc committed:
refactor: 重命名

- 将包名从 "y-mongodb-provider" 改为 "y-mongodb-provider-crashsafe"
- 更新 GitHub仓库 URL 和 issue 跟踪链接
- 修改包的 homepage URL

lukenc committed:
fix(y-mongodb.js): Correct off-by-one error in update logic

- In `src/y-mongodb.js`:
  - Fixed an issue where under certain conditions documents were not flushed correctly. The previous condition `applyNum === updates.length - 1` did not ensure that a flush would occur when `updates.length` was exactly equal to `this.flushSize`. The condition has been corrected to `applyNum === updates.length` to ensure documents are flushed after all updates are applied.

Additionally, modifications in `package-lock.json` include:
- Renaming the project from `y-mongodb-provider` to `y-mongodb-provider-crashsafe` to reflect enhanced crash safety.

lukenc committed:
fix(y-mongodb.js): Ensure correct flush behavior and simplify warning messages

- In `src/y-mongodb.js`:
  - Corrected the condition for flushing documents after applying updates. Now ensures that a flush occurs only when all updates are successfully applied (`applyNum === updates.length`), addressing an off-by-one error in the previous logic.
  - Simplified the warning message when failing to apply an update by removing redundant information.

Additionally, modifications in `package.json` include:
- Updated the version from `0.2.0` to `0.2.2`, reflecting the bug fixes and improvements made in this commit.

MaxNoetzold commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
-	"name": "y-mongodb-provider",
-	"version": "0.2.0",
+	"name": "y-mongodb-provider-crashsafe",
```
I wont do that

MaxNoetzold commented on package.json:
```diff
@@ -23,12 +23,12 @@
 	"repository": {
 		"type": "git",
-		"url": "git+https://github.com/MaxNoetzold/y-mongodb-provider.git"
+		"url": "git+https://github.com/lukenc/y-mongodb-provider.git"
```
nor this

MaxNoetzold commented on package.json:
```diff
@@ -23,12 +23,12 @@
 	"license": "MIT",
 	"bugs": {
-		"url": "https://github.com/MaxNoetzold/y-mongodb-provider/issues"
+		"url": "https://github.com/lukenc/y-mongodb-provider/issues"
```
nor that

MaxNoetzold commented on package.json:
```diff
@@ -54,7 +54,7 @@
 		"src/*"
 	],
-	"homepage": "https://github.com/MaxNoetzold/y-mongodb-provider#readme",
+	"homepage": "https://github.com/lukenc/y-mongodb-provider#readme",
```
same

MaxNoetzold commented on src/y-mongodb.js:
```diff
@@ -101,13 +101,27 @@ export class MongodbPersistence {
 		return this._transact(docName, async (db) => {
 			const updates = await U.getMongoUpdates(db, docName);
 			const ydoc = new Y.Doc();
+			let applyNum = 0;
```
this variable doesnt say what it does, it needs some kind of renaming

MaxNoetzold commented on src/y-mongodb.js:
```diff
@@ -101,13 +101,27 @@ export class MongodbPersistence {
-					Y.applyUpdate(ydoc, updates[i]);
+					try {
+						Y.applyUpdate(ydoc, updates[i]);
+					} catch (e) {
```
you catch all errors now and remove the option for error handling at later stages at all. This is probably a bad idea, but I am not super sure if error handling was an option anyways because of the transact wrapper. 

However, I am even wondering if this makes any sense. If an error is thrown when applying an update, the flush later on wont be happening without your changes, so it doesnt change anything but just adds logging?

MaxNoetzold commented on src/y-mongodb.js:
```diff
@@ -101,13 +101,27 @@ export class MongodbPersistence {
 			});
-			if (updates.length > this.flushSize) {
+			if (updates.length > this.flushSize && applyNum === updates.length) {
+				// 情况1: 更新数量超过阈值并且全部应用成功，执行文档刷新
```
I only want english comments here, sorry

MaxNoetzold commented on src/y-mongodb.js:
```diff
@@ -101,13 +101,27 @@ export class MongodbPersistence {
 				}
 			});
-			if (updates.length > this.flushSize) {
+			if (updates.length > this.flushSize && applyNum === updates.length) {
+				// 情况1: 更新数量超过阈值并且全部应用成功，执行文档刷新
 				await U.flushDocument(db, docName, Y.encodeStateAsUpdate(ydoc), Y.encodeStateVector(ydoc));
+			} else if (applyNum === updates.length) {
+				/* empty */
+			} else if (applyNum !== updates.length) {
+				// 情况3: 无法应用所有更新，记录警告
+				console.warn(
+					`Failed to apply all updates to document "${docName}". Applied ${applyNum}/${updates.length} updates.`,
+				);
 			}
```
These if/else blocks can be easily simplified:

```suggestion
			if (updates.length > this.flushSize && ) {
				if (applyNum === updates.length) {
					await U.flushDocument(db, docName, Y.encodeStateAsUpdate(ydoc), Y.encodeStateVector(ydoc));
				} else {
					console.warn(
					`Failed to apply all updates to document "${docName}". Applied ${applyNum}/${updates.length} updates.`,
				);
				}
			}
```

MaxNoetzold reviewed (changes requested):
Thank you for wanting to improve this package. I do see some points where this can be improved, see the comments. And you would also need to add some new tests for the new behavior.

lukenc committed:
fix(y-mongodb.js): Ensure correct flush behavior and simplify warning messages

- In `src/y-mongodb.js`:
  - Corrected the condition for flushing documents after applying updates. Now ensures that a flush occurs only when all updates are successfully applied (`applyNum === updates.length`), addressing an off-by-one error in the previous logic.
  - Simplified the warning message when failing to apply an update by removing redundant information.

Additionally, modifications in `package.json` include:
- Updated the version from `0.2.0` to `0.2.2`, reflecting the bug fixes and improvements made in this commit.

lukenc committed:
Add index creation for MongoDB collection to optimize queries

- Implemented `ensureIndexes` method in `src/mongo-adapter.js` to create indexes on `docName` and `clock` fields for a specified collection.
- Updated `src/utils.js` to call `ensureIndexes` on first write, ensuring indexes are created only once.

lukenc committed:
build: 更新版本号至 0.2.4

- 在 package.json 中将版本号从 0.2.3 修改为 0.2.4
